File: GUILib/src/space/kiibou/event/MouseEventAction.java
package space.kiibou.event;

import processing.event.MouseEvent;

import java.util.EnumSet;
import java.util.HashMap;
import java.util.Map;

public enum MouseEventAction {
    PRESS(processing.event.MouseEvent.PRESS),
    RELEASE(processing.event.MouseEvent.RELEASE),
    CLICK(processing.event.MouseEvent.CLICK),
    DRAG(processing.event.MouseEvent.DRAG),
    MOVE(processing.event.MouseEvent.MOVE),
    WINDOW_ENTER(processing.event.MouseEvent.ENTER),
    WINDOW_EXIT(processing.event.MouseEvent.EXIT),
    WHEEL(processing.event.MouseEvent.WHEEL),
    ELEMENT_ENTER(-1),
    ELEMENT_EXIT(-2);

    private static final Map<Integer, MouseEventAction> mapper = new HashMap<>();

    static {
        for (MouseEventAction action : MouseEventAction.values()) {
            if (action.id >= 0)
                mapper.put(action.id, action);
        }
    }

    private int id;

    MouseEventAction(int id) {
        this.id = id;
    }

    public static EnumSet<MouseEventAction> fromProcessingEvent(MouseEvent event) {
        return EnumSet.of(mapper.get(event.getAction()));
    }
}

File: GUILib/src/space/kiibou/event/MouseEventAction.kt
package space.kiibou.event

import processing.event.MouseEvent
import java.util.*

enum class MouseEventAction(private val id: Int) {
    PRESS(MouseEvent.PRESS),
    RELEASE(MouseEvent.RELEASE),
    CLICK(MouseEvent.CLICK),
    DRAG(MouseEvent.DRAG),
    MOVE(MouseEvent.MOVE),
    WINDOW_ENTER(MouseEvent.ENTER),
    WINDOW_EXIT(MouseEvent.EXIT),
    WHEEL(MouseEvent.WHEEL),
    ELEMENT_ENTER(-1),
    ELEMENT_EXIT(-2);

    companion object {
        private val mapper: Map<Int, MouseEventAction> = values().filter { it.id > 0 }.map { it.id to it }.toMap()

        fun fromProcessingEvent(event: MouseEvent): EnumSet<MouseEventAction> = EnumSet.of(mapper[event.action])
    }

}